### PR TITLE
Modify web retrieval

### PR DESCRIPTION
**Pull request by dbobrenko**
## Changes
  - Modify web retrieval to make miners return URLs along with content.
  - Modify expected miner's response to be a JSON string with the fields: `url`, `content`, `relevant`.
  - Modify reward function to verify provided URL content.
  - Fix some type annotations.

### CODE REVIEW

File: prompting/rewards/web_retrieval.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+        # URL and the content provided in the completion.
+        response_url, response_content = self._parse_response(completion)
+        # Content scraped from the URL provided in the completion.
+        response_url_scraped = self._extract_website_content(response_url)
+
+        # Similarity between search term and miner's scraped content.
+        search_response_sim = self._cosine_similarity(content1=search_term, content2=response_content)
+        # Similarity between reference conent and miner's scraped content.
+        # reference_response_sim = self._cosine_similarity(content1=reference_content, content2=response_content)
```
Comment (Hollyqui) [inline prompting/rewards/web_retrieval.py:47]: remove comments unless we plan to uncomment

File: prompting/rewards/web_retrieval.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        Args:
+            completion: The text to parse.
+
+        Returns:
+            tuple: A tuple containing the URL (str or None) and the content (str).
+        """
+        url = None
+        lines = completion.strip().split("\n")
+
+        # First, try to find URL by parsing the last line that starts with "URL:".
```
Comment (Hollyqui) [inline prompting/rewards/web_retrieval.py:89]: should we really try to parse them or should we just make them return a json like

```
[
{
   "website_content": ...,
   "website_url": ...,
   "relevant_section": ...,
},
]
```

for instance? That way they can give us url, the most relevant section + the whole website content without us having to do any awkward parsing

**Review by Hollyqui — COMMENTED** [review]
Check comments, otherwise good to go in principle

File: prompting/rewards/web_retrieval.py (hunk 2 above)
Comment (dbobrenko) [reply prompting/rewards/web_retrieval.py:87]: There is no parsing as such, it just searches for URL and the end of response
However we can do JSON approach, which is less error prone

**Review by bkb2135 — APPROVED** [review]

**Pull request merged**